crates/phactory/src/contracts/geolocation.rs

```diff
@@ -73,18 +83,46 @@ impl contracts::NativeContract for Geolocation {
         cmd: Command,
     ) -> TransactionResult {
         match cmd {
-            Command::UpdateGeolocation { geolocation_info } => {
+            Command::UpdateGeolocation { geocoding, err_info } => {
```

Comment:
Well done. That's a very complex function. However I think you may consider an easier option:

1. We have `geo_data` mapping from the worker pubkey to its latest submission. On each newly arrived command, we update the data with trivial effort.
2. We add another handler executed after the process of each block. In this handler, we first remove the expired data from `geo_data`, and then rebuild the reverse map `region_map`.
3. pRuntime handles requests in single thread mode. So you don't need to worry user's query will arrive while it's still processing the commands

Reply:
To @kvinwang @shelvenzhou: Do we have some hook functions like `on_block_end` that can be executed at every block? If not, I think it makes sense to add one.